crypto/zk/artifacts/artifacts.go

```diff
@@ -0,0 +1,175 @@
+	// LocalDir defines in which directory will be the files
+	// downloaded, under that directory it will follow the CircuitsPath
+	// directories structure
+	LocalDir string
+
+	// WitnessHash contains the expected hash for the file filenameWitness
+	WitnessHash []byte // witness.wasm
+	// ZkeyHash contains the expected hash for the file filenameZkey
+	ZkeyHash []byte // circuit.zkey
+	// VkHash contains the expected hash for the file filenameVk
+	VkHash []byte // verificationKey.json
+}
+
+// DownloadCircuitFiles will download the circuits in the specified path,
+// checking the expected sha256 hash of each file. If the files already exist
+// in the path, it will not download them again but it will check the hash of
+// the existing files.
+func DownloadCircuitFiles(c CircuitsConfig) error {
+	// check if files already exist, if files already exist, check hashes
+	err := checkHashes(c)
+	if err == nil {
+		// files already exist, and match the expected hashes
+		return nil
+	}
+	if err != nil && !strings.Contains(err.Error(), "no such file or directory") {
+		return err
+	}
+
+	// at this point, or files do not exist, or exist with a not expected
+	// hash
+	return downloadFiles(c)
+}
+
+func checkHashes(c CircuitsConfig) error {
+	err := checkHash(c.LocalDir+c.CircuitsPath+FilenameWitness, c.WitnessHash)
+	if err != nil {
+		return err
+	}
+
+	err = checkHash(c.LocalDir+c.CircuitsPath+FilenameZkey, c.ZkeyHash)
+	if err != nil {
+		return err
+	}
+
+	err = checkHash(c.LocalDir+c.CircuitsPath+FilenameVk, c.VkHash)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func checkHash(path string, expected []byte) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return err
+	}
+	hash := h.Sum(nil)
+
+	if !bytes.Equal(hash[:], expected) {
+		return fmt.Errorf("path: %s, expected hash: %s, computed hash: %s",
+			path, hex.EncodeToString(expected), hex.EncodeToString(hash[:]))
+	}
+	return nil
+}
+
+// downloadFiles downloads the files, and once downloaded calls checkHashes
+func downloadFiles(c CircuitsConfig) error {
+	err := os.MkdirAll(c.LocalDir+c.CircuitsPath, os.ModePerm)
+	if err != nil {
+		return err
+	}
+	err = downloadFile(c.URL+c.CircuitsPath+FilenameWitness,
+		c.LocalDir+c.CircuitsPath+FilenameWitness)
+	if err != nil {
+		return err
+	}
+
+	err = downloadFile(c.URL+c.CircuitsPath+FilenameZkey,
+		c.LocalDir+c.CircuitsPath+FilenameZkey)
+	if err != nil {
+		return err
+	}
+
+	err = downloadFile(c.URL+c.CircuitsPath+FilenameVk,
+		c.LocalDir+c.CircuitsPath+FilenameVk)
+	if err != nil {
+		return err
+	}
+
+	err = checkHashes(c)
+	if err != nil {
+		return fmt.Errorf("error on download files from %s, %s", c.URL, err)
+	}
+
+	return nil
+}
+
+func downloadFile(fromUrl, toPath string) error {
+	out, err := os.Create(toPath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	resp, err := http.Get(fromUrl)
```

Comment:
I think it would be much better if we can have a timeout (maybe 1 minute?  Or a reasonable value for the expected size of the files, we can adjust it in the future) for the http request.

I think this should work https://stackoverflow.com/a/25344458

Another option would be to pass a context and decide the timeout somewhere else.  Like this:
```go
req, _ := http.NewRequest("GET", fromUrl, nil)
req = req.WithContext(ctx)
res, error := client.Do(req)
```
And somewhere else there would be:
```go
ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
defer cancel()
downloadFile(ctx, fromUrl, toPath)
```

Reply:
Cool! Done (using directly `http.NewRequestWithContext` https://pkg.go.dev/net/http#NewRequestWithContext )